src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -158,6 +182,823 @@ MultiLogicalPlanCreate(Query *originalQuery, Query *queryTree)
+ *   (i) subquery pushdown
+ *   (ii) single relation repartition subquery
+ *
+ * In order to create the logical plan, we follow the algorithm below:
+ *    -  If subquery pushdown planner can plan the query
+ *        -  We're done, we create the multi plan tree and return
+ *    -  Else
+ *       - If the query is not eligible for single table repartition subquery planning
+ *            - Throw the error that the subquery pushdown planner generated
+ *       - If it is eligible for single table repartition subquery planning
+ *            - Check for the errors for single table repartition subquery planning
+ *                - If no errors found, we're done. Create the multi plan and return
+ *                - If found errors, throw it
+ */
+static MultiNode *
+MultiSubqueryPlanTree(Query *originalQuery, Query *queryTree,
+					  PlannerRestrictionContext *plannerRestrictionContext)
+{
+	MultiNode *multiQueryNode = NULL;
+	DeferredErrorMessage *subqueryPushdownError = NULL;
+
+	/*
+	 * This is a generic error check that applies to both subquery pushdown
+	 * and single table repartition subquery.
+	 */
+	ErrorIfQueryNotSupported(originalQuery);
+
+	/*
+	 * In principle, we're first trying subquery pushdown planner. If it fails
+	 * to create a logical plan, continue with trying the single table
+	 * repartition subquery planning.
+	 */
+	subqueryPushdownError = DeferErrorIfContainsUnsupportedSubqueryPushdown(originalQuery,
+																			plannerRestrictionContext);
+	if (!subqueryPushdownError)
+	{
+		multiQueryNode = SubqueryPushdownMultiPlanTree(originalQuery);
+	}
+	else if (subqueryPushdownError)
+	{
+		bool singleRelationRepartitionSubquery =
+			SingleRelationRepartitionSubquery(originalQuery);
+
+		RangeTblEntry *subqueryRangeTableEntry = NULL;
+		Query *subqueryTree = NULL;
+		DeferredErrorMessage *repartitionQueryError = NULL;
+		List *subqueryEntryList = NULL;
+
+		/*
+		 * If not eligible for single relation repartition query, we should raise
+		 * subquery pushdown error.
+		 */
+		if (!singleRelationRepartitionSubquery)
+		{
+			RaiseDeferredErrorInternal(subqueryPushdownError, ERROR);
+		}
+
+		subqueryEntryList = SubqueryEntryList(queryTree);
+		subqueryRangeTableEntry = (RangeTblEntry *) linitial(subqueryEntryList);
+		Assert(subqueryRangeTableEntry->rtekind == RTE_SUBQUERY);
+
+		subqueryTree = subqueryRangeTableEntry->subquery;
+
+		repartitionQueryError = DeferErrorIfSubqueryNotSupported(subqueryTree);
+		if (repartitionQueryError)
+		{
+			RaiseDeferredErrorInternal(repartitionQueryError, ERROR);
+		}
+
+		/* all checks has passed, safe to create the multi plan */
+		multiQueryNode = MultiPlanTree(queryTree);
+	}
+
+	Assert(multiQueryNode != NULL);
+
+	return multiQueryNode;
+}
+
+
+/*
+ * SingleRelationRepartitionSubquery returns true if it is eligible single
+ * repartition query planning in the sense that:
+ *   - None of the levels of the subquery contains a join
+ *   - Only a single RTE_RELATION exists
```

Comment:
Oh, sorry I misunderstood your comment. This function checks for 3 things (comment updated in the PR as well): 

1. No joins at all
2. Only a single `RTE_RELATION`, which means only a single table name is specified on the whole query 
3. No subqueries in WHERE clause

> I meant "Only a single RTE_RELATION exists" should deduce that we don't have joins. I'm okay to keep the first line, just noting to see I missed anything?

In fact, that comment is intended for the second item in the list. Is that better now?